src/model/encoding/convertFromRawToDraftState.js

```diff
@@ -13,60 +13,229 @@
 
-    var inlineStyles = decodeInlineStyleRanges(text, inlineStyleRanges);
+          if (!isValidBlock) {
+            invariant(
+              isValidBlock,
+              'invalid RawDraftContentBlock can not be converted to ContentBlockNode',
+            );
+            break;
+          }
 
-    // Translate entity range keys to the DraftEntity map.
-    var filteredEntityRanges = entityRanges
-      .filter(range => fromStorageToLocal.hasOwnProperty(range.key))
-      .map(range => {
-        return {...range, key: fromStorageToLocal[range.key]};
-      });
+          // ensure children have valid keys to enable sibling links
+          const children = node.children.map(addKeyIfMissing);
 
-    var entities = decodeEntityRanges(text, filteredEntityRanges);
-    var characterList = createCharacterList(inlineStyles, entities);
+          const contentBlockNode = new ContentBlockNode({
+            ...decodeBlockNodeConfig(node, entityMap),
+            parent: parentRef.getKey(),
+            children: List(children.map(child => child.key)),
+            prevSibling: index === 0 ? null : siblings.get(index - 1),
+            nextSibling:
+              index === siblings.size - 1 ? null : siblings.get(index + 1),
+          });
 
-    return new ContentBlock({key, type, text, depth, characterList, data});
+          // push node to blockMap
+          acc.push(contentBlockNode);
+
+          // this stack is used to ensure we visit all nodes respecting depth ordering
+          stack = updateNodeStack(stack, children, contentBlockNode);
+        }
+
+        return acc;
+      }, [])
+  );
+};
+
+const decodeContentBlocks = (
+  blocks: Array<RawDraftContentBlock>,
+  entityMap: *,
+): BlockMap => {
+  return blocks.map(
+    (block: RawDraftContentBlock) =>
+      new ContentBlock(decodeBlockNodeConfig(block, entityMap)),
+  );
+};
+
+const decodeRawBlocks = (
+  rawState: RawDraftContentState,
+  entityMap: *,
+): BlockMap => {
+  const rawBlocks = rawState.blocks;
+
+  if (!Array.isArray(rawBlocks[0].children)) {
+    return decodeContentBlocks(rawBlocks, entityMap);
+  }
+
+  return decodeContentBlockNodes(rawBlocks, entityMap);
+};
+
+const decodeRawEntityMap = (rawState: RawDraftContentState): * => {
+  const {entityMap: rawEntityMap} = rawState;
+  const entityMap = {};
+
+  // TODO: Update this once we completely remove DraftEntity
+  Object.keys(rawEntityMap).forEach(rawEntityKey => {
+    const {type, mutability, data} = rawEntityMap[rawEntityKey];
+
+    // get the key reference to created entity
+    entityMap[rawEntityKey] = DraftEntity.__create(
+      type,
+      mutability,
+      data || {},
+    );
   });
 
-  return ContentState.createFromBlockArray(contentBlocks);
-}
+  return entityMap;
+};
+
+const convertFromRawToDraftState = (
+  rawState: RawDraftContentState,
+): ContentState => {
+  invariant(Array.isArray(rawState.blocks), 'invalid RawDraftContentState');
+
```

Comment:
I really like this new refactor where we split `decodeRawEntityMap` and `decodeRawBlocks` into separate functions.